lib/translator/report_luns.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "report_luns.h"
+
+#include "absl/base/casts.h"
+
+namespace translator {
+
+namespace {
+
+// Return the number of valid namespace IDs in a namespace ID list.
+uint32_t GetNsListLength(const nvme_defs::IdentifyNamespaceList& ns_list) {
+  uint32_t size = 0;
+  for (uint32_t i = 0; i < nvme_defs::kIdentifyNsListMaxLength; ++i) {
+    if (ns_list.ids[i] == 0) break;
+    ++size;
+  }
+  return size;
+}
+
+}  // namespace
+
+// Section 4.5
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToNvme(absl::Span<const uint8_t> scsi_cmd,
+                            nvme_defs::GenericQueueEntryCmd& identify_cmd,
+                            uint32_t& alloc_len) {
+  if (scsi_cmd.size() != sizeof(scsi_defs::ReportLunsCommand)) {
+    DebugLog("Malformed ReportLuns command");
+    return StatusCode::kInvalidInput;
+  }
+
+  // Cast scsi_cmd to ReportLunsCommand
+  scsi_defs::ReportLunsCommand rl_cmd;
+  ReadValue(scsi_cmd, rl_cmd);
+  if (rl_cmd.select_report != scsi_defs::SelectReport::kRestrictedMethods &&
+      rl_cmd.select_report != scsi_defs::SelectReport::kWellKnown &&
+      rl_cmd.select_report != scsi_defs::SelectReport::kAllLogical) {
+    DebugLog("Invalid report luns select report %u", rl_cmd.select_report);
+    return StatusCode::kInvalidInput;
+  }
+
+  // Assign allocation length for downstream use
+  alloc_len = rl_cmd.alloc_length;
+
+  // Assign generic params to NVMe command
+  identify_cmd = {};
+  identify_cmd.opc = static_cast<uint8_t>(nvme_defs::AdminOpcode::kIdentify);
+  identify_cmd.cdw[0] = 0x2;  // Set CNS to return namespace ID list
```

Comment:
for a CNS value of 0x2, I think the needs to be set. Figure 244 https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf

Reply:
In this case the NSID should be 0. Since the NVMe structure is 0 initialized when identify_cmd = {}, I neglected setting it redundantly. If the general consensus is to explicitly set the NSID to 0 we can do that, but imo it's cleaner to just 0 initialize